### PR TITLE
Initial stab at #251

### PR DESCRIPTION
**Pull request by joewalker**
Needs rebase onto `rust`

### CODE REVIEW

File: tx-parser/Cargo.toml [hunk 1]
```diff
@@ -10,3 +10,9 @@ combine = "2.1.1"
 
 [dependencies.mentat_tx]
   path = "../tx"
+
+[dependencies.mentat_parser_utils]
+  path = "../parser-utils"
+
+[dependencies.mentat_query_parser]
```
Comment (ncalexan) [inline tx-parser/Cargo.toml]: I don't expect a dependency between parsing transactions and parsing queries.  Anything that's shared is really about parsing `edn::Value`, I expect, and should live in the parsing utilities.

File: tx-parser/src/lib.rs [hunk 2]
```diff
@@ -13,12 +13,17 @@
 extern crate edn;
 extern crate combine;
 extern crate mentat_tx;
+extern crate mentat_parser_utils;
+
+#[macro_use]
+extern crate mentat_query_parser;
```
Comment (ncalexan) [inline tx-parser/src/lib.rs]: Further to the above, I expect the macros to come from the parsing utilities crate.

File: tx-parser/src/lib.rs [hunk 3]
```diff
@@ -30,9 +35,19 @@ fn fn_parser<O, I>(f: fn(I) -> ParseResult<O, I>, err: &'static str) -> TxParser
...
+        } else {
+            None
+        })
+        .parse_stream(input);
+});
+
 impl<I> Tx<I>
     where I: Stream<Item = Value>
 {
+    /*
```
Comment (ncalexan) [inline tx-parser/src/lib.rs]: If this is really unused, please kill it for good.

File: parser-utils/src/query.rs [hunk 4]
```diff
@@ -11,7 +11,6 @@
 extern crate combine;
```
Comment (ncalexan) [inline parser-utils/src/query.rs]: These utilities aren't really query specific, they're `edn::Value` specific.  (Am I wrong?)  Can we move to `src/value_parsing.rs` or something you prefer?

File: parser-utils/Cargo.toml [hunk 5]
```diff
@@ -5,3 +5,9 @@ authors = ["Victor Porof <vporof@mozilla.com>", "Richard Newman <rnewman@mozilla
 
 [dependencies]
 combine = "2.1.1"
+
+[dependencies.edn]
+  path = "../edn"
+
+[dependencies.mentat_query]
```
Comment (ncalexan) [inline parser-utils/Cargo.toml]: Can you explain why parsing utilities would depend on the structure of queries?  All the parsing utilities are likely to be abstract over their result type.

File: parser-utils/src/query.rs (hunk 4 above)
Comment (rnewman) [inline parser-utils/src/query.rs]: Or, indeed, just into the `lib.rs`. They're all parser utils.

File: tx-parser/src/lib.rs [hunk 6]
```diff
@@ -30,9 +35,19 @@ fn fn_parser<O, I>(f: fn(I) -> ParseResult<O, I>, err: &'static str) -> TxParser
     parser(f).expected(err)
 }
 
+def_parser_fn!(Tx, integer, Value, i64, input, {
+    return satisfy_map(|x: Value| if let Value::Integer(y) = x {
```
Comment (rnewman) [inline tx-parser/src/lib.rs]: This can be phrased in terms of `def_value_satisfy_parser_fn!`.

**Comment by joewalker** [comment]
My intent with this PR was that it address #251 (the global imports one, which Jordan has now anyway) but the PR contains the changes in the branch on which this was based. So I'll close this, and address the comments in https://github.com/mozilla/mentat/pull/249

File: tx-parser/Cargo.toml (hunk 1 above)
Comment (joewalker) [inline tx-parser/Cargo.toml]: Yes, it wasn't needed.

File: tx-parser/src/lib.rs (hunk 3 above)
Comment (joewalker) [inline tx-parser/src/lib.rs]: Yup. Pushed mid hack. It's gone now.

File: parser-utils/src/query.rs (hunk 4 above)
Comment (joewalker) [inline parser-utils/src/query.rs]: Moved to lib.rs, my logic being that `matches_plain_symbol` uses `satisfy_unwrap` so it's good to have it close by.

File: parser-utils/Cargo.toml (hunk 5 above)
Comment (joewalker) [inline parser-utils/Cargo.toml]: They don't. Fixed.

**Pull request closed without merging**